### ISSUE TITLE
NVIDIA-01D Reference Cross-Check — midpoint flow warp / synthesis

### ISSUE THREAD

**Issue by piuteu** [piuteu/OBS-FrameGeneration-NVIDIA#9]
## Purpose

Record the small set of sources that directly inform NVIDIA-01D midpoint synthesis. These are used to prevent sign/scale/warp mistakes, not to import another frame-generation architecture.

## Source priority

1. NVIDIA primary documentation relevant to the NVOFA flow representation already established in 01C
2. Peer-reviewed interpolation math for constructing midpoint-to-input flow from bidirectional endpoint flow
3. Permissively licensed implementation references used only to cross-check backward-warp conventions
4. Other frame-interpolation/splatting projects only if the minimal baseline exposes a concrete problem

The existing NVIDIA-01 project architecture remains the source of truth. No reference below authorizes old-build reuse, NvOFFRUC integration, learned refinement, or scope expansion.

---

## A. NVIDIA DALI optical-flow documentation — grid-to-full-resolution scale rule

Relevant official NVIDIA documentation:

`nvidia.dali.fn.optical_flow`

Critical fact for 01D:

- NVOFA output at grid 4 is spatially coarser: one vector corresponds to each 4x4 input region;
- when resampling a full-resolution image using `output_grid=4`, NVIDIA explicitly states that the flow field is upsampled **without scaling the vector quantities**.

Implementation consequence:

- decode S10.5 with `/ 32.0`;
- reconstruct/sample the coarse flow field at full-resolution coordinates;
- do **not** multiply the displacement vector by 4 merely because the vector grid is 4x4.

This rule is authoritative for 01D vector magnitude handling.

---

## B. NVIDIA Optical Flow FRUC Programming Guide — high-level architecture confirmation only

Relevant official NVIDIA documentation:

`NVOFA FRUC Programming Guide`

Useful high-level fact:

- NVIDIA's FRUC description uses consecutive frames;
- obtains forward and backward optical flow between previous/next frames;
- then uses CUDA-accelerated techniques to generate an interpolated frame;
- the interpolated frame is interleaved with original frames.

This confirms that our high-level `A/B -> FWD+BWD -> CUDA synthesis -> real/generated sequence` direction is reasonable.

**Do not implement or link NvOFFRUC.** NvOFFRUC remains excluded by AD-003. The FRUC guide is architecture/background evidence only; its internal synthesis algorithm is not treated as disclosed or inferred.

---

## C. Super SloMo (CVPR 2018) — midpoint-flow approximation and backward-warp math

Paper:

H. Jiang et al., `Super SloMo: High Quality Estimation of Multiple Intermediate Frames for Video Interpolation`, CVPR 2018.

The paper defines endpoint flows `F0->1` and `F1->0` and approximates the required intermediate-to-endpoint flows as:

```text
F_t->0 = -(1-t)t * F_0->1 + t^2 * F_1->0
F_t->1 = (1-t)^2 * F_0->1 - t(1-t) * F_1->0
```

For our fixed midpoint `t = 0.5`:

```text
F_mid->A = -0.25 * F_A->B + 0.25 * F_B->A
F_mid->B =  0.25 * F_A->B - 0.25 * F_B->A
```

The paper then synthesizes the intermediate image by backward warping the two endpoint images with bilinear sampling and combining the warped results. It explicitly notes that the simple approximation works well in locally smooth regions but produces artifacts around motion boundaries; its learned refinement/visibility network addresses those artifacts.

01D consequence:

- the paper's **non-learned approximation** is a good minimal foundation baseline;
- do not import the refinement network or visibility model;
- do not make motion-boundary artifacts a foundation blocker;
- classify such artifacts for later quality work.

This paper is a mathematical reference. Implement independently; do not copy third-party code.

---

## D. Google FILM utility — backward-warp convention cross-check

Repository:

`google-research/frame-interpolation`

Relevant file:

`models/film_net/util.py`

License: Apache-2.0.

Useful cross-check:

- its `warp(image, flow)` documentation defines backward warping as target pixel `(x,y)` sampling the source image at `(x + flow_x, y + flow_y)` with bilinear lookup;
- flow channel order is `(dx, dy)`.

Use this only to cross-check coordinate/sign convention. 01D should implement a small native CUDA kernel suited to our persistent GPU-resident buffers rather than porting TensorFlow code.

---

## E. Forward splatting references — explicitly deferred

Projects/papers such as FILM's forward-warp/splatting path and M2M-VFI demonstrate why forward splatting can be valuable for holes/occlusion and motion boundaries.

For NVIDIA-01D these are **not baseline requirements**.

Do not add atomic splatting, softmax splatting, hole filling, depth weighting, learned visibility, or a hybrid warp/splat pipeline unless the minimal backward-warp midpoint cannot satisfy the basic correctness proof.

If the minimal baseline works but shows expected boundary/occlusion artifacts, record them and move on. Those artifacts belong to post-foundation quality refinement.

---

## Required Codex cross-check before implementation

Before finalizing the 01D CUDA kernel, compare the planned math and resource path against this issue and record in the PR:

1. S10.5 decode (`/32`);
2. 4x4 flow-grid reconstruction and the **no vector x4 scaling** rule;
3. A/B and forward/backward signs;
4. `t=0.5` intermediate-flow formula;
5. backward-warp sample convention (`source = target + intermediate_flow`);
6. bilinear image sampling / edge validity;
7. which existing 01C input slots remain owned until synthesis is complete;
8. where generated RGBA is stored before D3D11 presentation;
9. synchronization from NVOFA output -> synthesis -> generated-output-ready;
10. why forward splatting / learned refinement was not added to the baseline.

The goal is to prove a clean midpoint foundation, not to chase final visual quality.
